db/swappable_db_test.go

```diff
@@ -164,3 +172,55 @@ func Test_SwapInvalidSQLiteFile(t *testing.T) {
 		t.Fatalf("expected an error when swapping with an invalid SQLite file, got nil")
 	}
 }
+
+// Test_SwapRecoveryOnFailure tests that the Swap function recovers the original database
+// when the swap operation fails after the old database has been moved.
+func Test_SwapRecoveryOnFailure(t *testing.T) {
+	// Create original database with some data using regular DB
+	originalPath := mustTempPath()
+	defer os.Remove(originalPath)
+
+	// Now open it as SwappableDB
+	swappableDB, err := OpenSwappable(originalPath, nil, true, true)
+	if err != nil {
+		t.Fatalf("failed to open swappable database: %s", err)
+	}
+	mustExecute(swappableDB.db, "CREATE TABLE original (id INTEGER NOT NULL PRIMARY KEY, name TEXT)")
+	mustExecute(swappableDB.db, `INSERT INTO original(name) VALUES("original_data")`)
+	defer swappableDB.Close()
+
+	// Create a file that would pass the IsValidSQLiteFile check but fail when opened
+	swapPath := mustTempPath()
+	defer os.Remove(swapPath)
+
+	err = syscall.Mkfifo(swapPath, 0666)
```

Comment:
If we modify Swap so that it always leaves the original database place if the swap fails then we can remove the SQLite validity check from the implementation. That was only there to try to protect us because Swap() was a one-way operation. But if we fix that we can remove the check.